Fabric port

embeddedt opened the pull request:
Here you go!

Notable changes from the Forge version:

* I had to refactor the module logic to get rid of capabilities. Now, an `Item` can implement the `ScannerModuleProvider` interface to be treated as a module. I haven't tested this in a custom mod but it should work since the mod uses that interface even for its own items.
* Instead of using the Forge tags, I'm using the Fabric-style ore tags (e.g. `c:coal_ores`).
* The concept of storing inactive modules inside the scanner has been removed, since I had to reimplement the inventory logic myself on Fabric and couldn't get it to work.

Aside from that, Mojmap + Architectury make this a very straightforward port overall.

Let me know if you think anything needs to be changed.

embeddedt committed:
Compiling on Fabric 1.18.1

embeddedt committed:
More functionality works

embeddedt committed:
Get OverlayRenderer working

embeddedt committed:
Basic scanning works

embeddedt committed:
Misc polishing

embeddedt committed:
Remove inactive module slots

embeddedt committed:
Refactor ScannerModuleProvider to pass an ItemStack

embeddedt committed:
Update README

embeddedt committed:
Port to 1.18.2

embeddedt committed:
Merge remote-tracking branch 'origin/1.18-fabric' into 1.18-fabric

embeddedt committed:
Restore CurseGradle, etc.

embeddedt committed:
Track data-generated files in Git

fnuecke commented on build.gradle:
```diff
@@ -33,99 +22,93 @@ def getGitRef() {
-    runs {
-        client {
-            workingDirectory project.file('run')
+    modImplementation "curse.maven:forgeconfigapi-547434:3671141"
+    implementation 'com.electronwill.night-config:core:3.6.3'
+    implementation 'com.electronwill.night-config:toml:3.6.3'
```
Are these shipped by Fabric? If not, could they be added as dependencies in the curse publishing task? (`curseforge`)

embeddedt replied:
Thanks for reminding me; I need to add both Forge Config API and Architectury as dependencies. Will get to that soon(ish).

fnuecke commented on src/main/java/li/cil/scannable/client/ScannableClient.java:
```diff
@@ -0,0 +1,31 @@
+        MenuRegistry.registerScreenFactory(Containers.BLOCK_MODULE_CONTAINER.get(), ConfigurableBlockScannerModuleContainerScreen::new);
+        MenuRegistry.registerScreenFactory(Containers.ENTITY_MODULE_CONTAINER.get(), ConfigurableEntityScannerModuleContainerScreen::new);
+
+        //WorldRenderEvents.BEFORE_DEBUG_RENDER.register(OverlayRenderer::onOverlayRender);
```
WIP or can this be removed?

embeddedt replied:
Yes, this can be removed. I ended up using a mixin for that renderer because Fabric didn't have a good hook.

fnuecke commented on src/main/java/li/cil/scannable/client/shader/Shaders.java:
```diff
@@ -99,7 +103,8 @@ public void reload(final ResourceProvider provider) {
                 }, name, format);
             } catch (final Exception e) {
-                LOGGER.error(e);
+                e.printStackTrace();
+                //LOGGER.error(e);
```
Was this for testing and can be reverted?

embeddedt replied:
Yes; I remember making this change while getting the shaders to load on Fabric. It's been reverted now.

fnuecke commented on src/main/java/li/cil/scannable/common/Scannable.java:
```diff
@@ -1,38 +1,25 @@
-        FMLJavaModLoadingContext.get().getModEventBus().register(CommonSetup.class);
-        DistExecutor.unsafeRunWhenOn(Dist.CLIENT, () -> () -> FMLJavaModLoadingContext.get().getModEventBus().register(ClientSetup.class));
+        Network.initialize();
+        //DistExecutor.unsafeRunWhenOn(Dist.CLIENT, () -> () -> FMLJavaModLoadingContext.get().getModEventBus().register(ClientSetup.class));
```
WIP or can this be removed?

embeddedt replied:
Removed; I commented it out while porting.

fnuecke commented on src/main/java/li/cil/scannable/common/item/ScannerItem.java:
```diff
@@ -178,16 +164,18 @@ public ItemStack finishUsingItem(final ItemStack stack, final Level level, final
             return stack;
         }
 
+        /*
         if (level.isClientSide()) {
             SoundCanceler.cancelEquipSound();
         }
+         */
```
Is this workaround not (no longer?) needed, or did it not work for Fabric?

embeddedt replied:
I'm not sure the equivalent event exists in Fabric (though a mixin could add it relatively easily), and I don't seem to hear any reequip sound or see anything in the subtitles when the scan finishes.

fnuecke replied:
Just tested a bit, and doesn't seem to happen anymore, so this can be completely removed 🎉

The issue this was added for was, that in non-creative, when energy stored changed, the NBT changed, the client was notified of that change, the client thought it was a new item, and played the equip sound. Looks like that's no longer necessary, so yay.

fnuecke commented on src/main/java/li/cil/scannable/common/scanning/FluidBlockScannerModule.java:
```diff
@@ -34,48 +32,37 @@ public int getEnergyCost(final ItemStack module) {
-
-    @OnlyIn(Dist.CLIENT)
-    @SubscribeEvent
-    public static void onModConfigEvent(final ModConfigEvent configEvent) {
```
Since the cache invalidation was ported for the common ores one, any reason it's not also done here?

embeddedt replied:
It was an oversight in both cases.

fnuecke commented on src/main/java/li/cil/scannable/common/scanning/RareOresBlockScannerModule.java:
```diff
@@ -40,61 +33,49 @@ public int getEnergyCost(final ItemStack module) {
-
-    @OnlyIn(Dist.CLIENT)
-    @SubscribeEvent
-    public static void onModConfigEvent(final ModConfigEvent configEvent) {
```
Ditto, any reason cache invalidation is not ported here?

fnuecke commented on src/main/java/li/cil/scannable/data/DataGenerators.java:
```diff
@@ -1,27 +1,15 @@
+        generator.addProvider(blockTagsProvider);
+        generator.addProvider(new ModItemTagsProvider(generator, blockTagsProvider));
+        generator.addProvider(new ModRecipeProvider(generator));
+        //generator.addProvider(new ModItemModelProvider(generator));
```
Can this be removed?

fnuecke commented on src/main/java/li/cil/scannable/data/ModItemModelProvider.java:
```diff
@@ -3,22 +3,21 @@
-public final class ModItemModelProvider extends ItemModelProvider {
-    public ModItemModelProvider(final DataGenerator generator, final ExistingFileHelper existingFileHelper) {
-        super(generator, API.MOD_ID, existingFileHelper);
+/*
```
WIP/fell between the cracks/unneeded?

embeddedt replied:
Model generation wasn't supported in Fabric at the time that I ported to 1.18.1. I think it is supported in 1.18.2 so I'll look at adding it back.

embeddedt replied:
I've added this back now.

fnuecke commented on src/main/java/li/cil/scannable/data/ModRecipeProvider.java:
```diff
@@ -8,55 +10,56 @@
-            .pattern("cnc")
-            .define('g', Tags.Items.DYES_GREEN)
-            .define('c', Items.CLAY_BALL)
-            .define('r', Tags.Items.DUSTS_GLOWSTONE)
-            .define('n', Tags.Items.NUGGETS_GOLD)
-            .group("blank_module")
-            .unlockedBy("has_scanner", InventoryChangeTrigger.TriggerInstance.hasItems(SCANNER.get()))
-            .save(consumer);
+                .pattern("ggg")
+                .pattern("crc")
+                .pattern("cnc")
+                .define('g', Items.GREEN_DYE)
+                .define('c', Items.CLAY_BALL)
+                .define('r', Items.GLOWSTONE_DUST)
+                .define('n', Items.GOLD_NUGGET)
+                .group("blank_module")
+                .unlockedBy("has_scanner", InventoryChangeTrigger.TriggerInstance.hasItems(SCANNER.get()))
+                .save(consumer);
 
-        registerModule(RANGE_MODULE.get(), Tags.Items.ENDER_PEARLS).save(consumer);
+        registerModule(RANGE_MODULE.get(), Items.ENDER_PEARL).save(consumer);
         registerModule(ENTITY_MODULE.get(), Items.LEAD).save(consumer);
-        registerModule(FRIENDLY_ENTITY_MODULE.get(), Tags.Items.LEATHER).save(consumer);
-        registerModule(HOSTILE_ENTITY_MODULE.get(), Tags.Items.BONES).save(consumer);
-        registerModule(BLOCK_MODULE.get(), Tags.Items.STONE).save(consumer);
+        registerModule(FRIENDLY_ENTITY_MODULE.get(), Items.LEATHER).save(consumer);
+        registerModule(HOSTILE_ENTITY_MODULE.get(), Items.BONE).save(consumer);
+        registerModule(BLOCK_MODULE.get(), Items.STONE).save(consumer);
         registerModule(COMMON_ORES_MODULE.get(), Items.COAL).save(consumer);
-        registerModule(RARE_ORES_MODULE.get(), Tags.Items.GEMS_DIAMOND).save(consumer);
+        registerModule(RARE_ORES_MODULE.get(), Items.DIAMOND).save(consumer);
```
No common tags for these in Fabric?

embeddedt replied:
The Fabric tag situation is a bit of a mess right now, in my opinion. They have some common tag *names*, but the API itself doesn't create or populate them. It's left up to mods to do that, even though it's redundant as soon as multiple are installed.

~~If you think it's worth adding data generation for the necessary common tags I'll do it.~~ In the process of adding back the item model datagen I decided to just add this too.

fnuecke commented on src/main/resources/scannable.mixins.json:
```diff
@@ -0,0 +1,14 @@
+    "required": true,
+    "minVersion": "0.8",
+    "package": "li.cil.scannable.mixin",
+    "compatibilityLevel": "JAVA_8",
```
Should this be JAVA_17 for 1.18.2+?

fnuecke reviewed (changes requested):
Awesome, thanks a bunch! Just a few minor questions here and there. Main change request (unless I missed something) is related to the cache invalidation on config change.

Could you give me a quick rundown on what the complications with the inactive module row were? Multiple ItemHandler equivalents in Fabric, UI/Container? For future reference.

Very interesting read, cool to see some of the small differences in API design (e.g. the Transaction stuff to simulate inventory interaction).

Planning to do some quick testing tomorrow or so.

fnuecke committed:
Use Java 17 in Github Actions.

fnuecke commented:
I've just updated the Github actions to use Java 17 on the Forge branch. If you rebase onto/merge that to get the pipelines running, that'd be great.

embeddedt committed:
Merge remote-tracking branch 'origin/1.18-forge' into 1.18-fabric

embeddedt committed:
Address a couple review comments

embeddedt committed:
Revert logger change for debugging

embeddedt committed:
Remove dead code

embeddedt committed:
Tweaks to data generation

embeddedt committed:
Use common tags in recipes

embeddedt committed:
Add Roughly Enough Items in dev environment

embeddedt committed:
Add cache invalidation to modules

embeddedt committed:
Prevent REI from bringing its own Fabric API

embeddedt committed:
Add Architectury and Forge Config API Port as CurseGradle deps

embeddedt commented:
@fnuecke Thanks for the comments; I believe I've addressed all your concerns.

> Could you give me a quick rundown on what the complications with the inactive module row were? Multiple ItemHandler equivalents in Fabric, UI/Container? For future reference.

I'd like to, but in all honesty, I really don't remember beyond the fact that the slots simply didn't function. :laughing: It's possible I just didn't find the right equivalent to what was in Forge. I assumed (possibly prematurely) that it wasn't an important feature, given that I've never used it, and the code was simpler without it. 

Inventories are one of the areas of Minecraft modding I am the least familiar with, so someone more experienced can probably get it to work easily.

fnuecke commented:
Do you know if there's an API in Fabric to wrap tooltips where necessary, to avoid them flowing out of the sceen? IIRC Forge patches in some logic to do this automatically.

![2022-04-18_14 00 42](https://user-images.githubusercontent.com/514903/163805799-52f0f84c-2b72-4913-a3e7-6247c7dca68d.png)

embeddedt commented:
@fnuecke On Fabric that bugfix is provided by an independent [mod](https://www.curseforge.com/minecraft/mc-mods/tooltipfix).

embeddedt committed:
Remove unneeded sound canceler logic

fnuecke commented:
Ah, neat. Would it make sense to depend on that mod, or is the general consensus to leave this up to the user/modpack maker?

embeddedt commented:
In the Fabric space it seems to be more conventional to leave this to the user, in order to give them more flexibility over exactly what parts of vanilla get tweaked.

fnuecke reviewed (approved):
Cool! LGTM, thanks again :)
